Update of data_encoder and tabular_encoder

TZTsai opened the pull request:
## Description
I have added a tutorial of TabDDPM now. However, at first, when I tried a larger dataset than Iris in the tutorial, such as a wine quality dataset, the plugin did not manage to generate data within the maximum number of trials. This was due to the preprocessing of the continous features. As of now, I have added a `FeatureEncoder` in "data_encoder.py", which is used to encode or transform a single column, and updated the `TabularEncoder` in "tabular_encoder.py", which now applies two types of `FeatureEncoder` on continuous features and discrete features respectively. This generalized version of `TabularEncoder` is compatible with the previous one, using Bayesian GMM to encode continuous features by default. It has successfully passed the previously written tests. Using a `TabularEncoder` with a "Gaussian quantile" `FeatureEncoder` on the continuous features, the `TabDDPM` plugin can successfully generate data in the tutorial without any retry.

## Checklist
- [X] I have followed the [Contribution Guidelines](https://github.com/vanderschaarlab/.github/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/vanderschaarlab/.github/blob/main/CODE_OF_CONDUCT.md)
- [X] I have commented my code following the [van der Schaar Lab Styleguide](https://github.com/vanderschaarlab/.github/blob/main/STYLEGUIDE.md)
- [X] I have labelled this PR with the relevant [Type labels](https://github.com/vanderschaarlab/.github/labels?q=Type%3A)
- [X] My changes are covered by tests

pravsels commented:
@TZTsai There was also another potential issue. If I train the model using a condition, I'm not able to generate without passing in a condition (which is the case for the other models). I think it's the way `use_label` is set depending on condition being passed in the `fit()` function of TabDDPM class, which causes it to error out in the `forward()` function of MLPDiffusion class.

TZTsai commented:
> @TZTsai There was also another potential issue. If I train the model using a condition, I'm not able to generate without passing in a condition (which is the case for the other models). I think it's the way `use_label` is set depending on condition being passed in the `fit()` function of TabDDPM class, which causes it to error out in the `forward()` function of MLPDiffusion class during generate flow.

Hello! As I understand, if `cond` is given, the model learns a conditional distribution $P(x|y)$. The to be able to generate samples of $x$, $P(y)$ should be somehow given in prior. For a classification task, I assume the prior distribution $P(y)$ is a multinormial distribution, which is learned from the training labels. During generation, I first randomly sample labels from this multinomial distribution and send them as the condition. However, for a regression task, how should I generate the condition if it is not given? Should I assume some prior distribution and learn it during training?

dionman commented:
I think there's also a bug in https://github.com/vanderschaarlab/synthcity/blob/fb1db01aedaddc351f3082f0fc151de39ab1c1fc/src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py#L953. You have to be minibatching the cond vector along with the data, otherwise this raises a dimension mismatch error

bcebere commented:
> > @TZTsai There was also another potential issue. If I train the model using a condition, I'm not able to generate without passing in a condition (which is the case for the other models). I think it's the way `use_label` is set depending on condition being passed in the `fit()` function of TabDDPM class, which causes it to error out in the `forward()` function of MLPDiffusion class during generate flow.
> 
> Hello! As I understand, if `cond` is given, the model learns a conditional distribution P(x|y). Then to be able to generate samples of x, P(y) should be somehow given in prior. For a classification task, I assume the prior distribution P(y) is a multinormial distribution, which is learned from the training labels. During generation, I first randomly sample labels from this multinomial distribution and send them as the condition. However, for a regression task, how should I generate the condition if it is not given? Should I assume some prior distribution and learn it during training?

@ZhaozhiQIAN Can you please help @TZTsai here?

TZTsai commented:
The problem with the previous TabDDPM is indeed the preprocessing of continuous features. In order to use Gaussian quantile transformation (as suggested by the original paper) within `TabularEncoder`, I added a `FeatureEncoder` in [data_encoder.py](https://github.com/vanderschaarlab/synthcity/blob/main/src/synthcity/plugins/core/models/data_encoder.py) and make it the baseclass of all column encoders or transformers, including BayesianGMM, GaussianQuantile, OneHotEncoder, StandardScalar, etc. I then updated [TabularEncoder](https://github.com/vanderschaarlab/synthcity/blob/424b41448dd389560a001cba484facedc2f12e90/src/synthcity/plugins/core/models/tabular_encoder.py#L154) to make it compatible with any type of FeatureEncoder. Please review the changes in these two files. 

I have run the tests in "test_tabular_encoder.py" and all of them have passed, except the one for `BinEncoder`. I updated `BinEncoder` as a subclass of `TabularEncoder`, because I think only a couple of methods are different between them. The test seems to assume the output columns are the same as the input ones, but as I found in the doc of `BinEncoder`, the discrete columns are encoded by OneHot. The the output columns should be more than the input ones, so I am a bit confused here.

In the end, After I use a custom `TabularEncoder` in the TabDDPM plugin, it successfully generated data without any retry in the tutorial :).

TZTsai commented:
> I think there's also a bug in
> 
> https://github.com/vanderschaarlab/synthcity/blob/fb1db01aedaddc351f3082f0fc151de39ab1c1fc/src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py#L953
> You have to be minibatching the cond vector along with the data, otherwise this raises a dimension mismatch error.

I have fixed this.

bcebere commented on setup.cfg:
```diff
@@ -34,8 +34,8 @@ python_requires = >=3.7
 install_requires =
     scikit-learn>=1.0
     nflows>=0.14
-    pandas>=1.3,<2.0
```
this will cause some crashes atm.

TZTsai replied:
Do you mean I should add `<2.0` at the end?

bcebere replied:
Yes, the tests will crash otherwise. `lifelines` doesn't support pandas2, and asdgan/ctgan crash for some reason from torch2

TZTsai replied:
I have updated

bcebere commented on tests/plugins/generic/test_ddpm.py:
```diff
@@ -19,11 +19,16 @@
+#     n_iter=1000,
+#     batch_size=1000,
+#     num_timesteps=30,
+#     model_type="tabnet",
```
Why is this commented?

TZTsai replied:
This was for testing the usage of TabNet as the diffusion model of TabDDPM. It fitted successfully but the performance was now as good as MLP.

bcebere reviewed (changes requested):
Great work! Just a few small observations

bcebere commented on setup.cfg:
```diff
@@ -58,7 +58,8 @@ install_requires =
     tsai; python_version>"3.7"
     importlib-metadata; python_version<"3.8"
-
+    igraph
```
Why are these in the default library depends?

TZTsai replied:
Sorry I should remove them, since they are in the extra requirements.

bcebere reviewed (changes requested):
Please fix the failing tests, and we will resume the review then.

bcebere reviewed (approved):
LGTM! 
@robsdavis please review as well.

floresdwm commented:
Very important note. 

data_encoder.py was removed after this update. Be aware after release 0.2.5, may produce this error of data_encoder.py module not found.
